Add instance actor (second take)

ClearlyClaire opened the pull request:
#10457 but using an actual `Account` with the following tricks:
- use `instance.tld` as username, to make sure we're not hitting an existing username
- use `Application` as `actor_type` and `0` as `id` to uniquely
- only serialize a subset of the `Actor` fields
- use its own route and controller to make it unique and avoid issues with dots in paths

It seems to work, I am not sure what may be missing but I haven't actually tested federating with it yet.

ClearlyClaire committed:
Add support for an instance actor

ClearlyClaire committed:
Skip username validation for local Application accounts

ClearlyClaire committed:
Add migration script to create instance actor

ClearlyClaire committed:
Make Codeclimate happy

ClearlyClaire committed:
Switch to id -99 for instance actor

Gargron commented on app/controllers/instance_actors_controller.rb:
```diff
@@ -0,0 +1,30 @@
+  end
+
+  def restrict_fields_to
+    %i(id type preferred_username inbox public_key endpoints icon image url manually_approves_followers)
```
What's the point of icon/image/url/manually_approves_followers?

ThibG replied:
`icon` and `image` were set to banner and hero or something like that in the other PR, but I have not done that, so it's not super useful, I'll remove them.

`manually_approves_followers` is kinda here to not confuse people when follows do not work

`url` is used to point at an explanation of what the actor is.

Gargron commented on app/serializers/activitypub/actor_serializer.rb:
```diff
@@ -39,10 +39,12 @@ def shared_inbox
   end
 
   def type
+    return 'Application' if object.instance_actor?
```
Feel this would be more readable in an if/elsif/else block.

ThibG replied:
Alternatively, I considered using `object.actor_type` directly. I think this *should* be fine, but I'm not completely sure all the contexts it is used in. Will do an if/elsif else block for now.

Gargron commented on app/views/instance_actors/show.html.haml:
```diff
@@ -0,0 +1,48 @@
+- content_for :page_title do
```
I can't really tell how this is going to look from code alone but do we *need* a HTML page for this?

ThibG replied:
The page is useful to let people know what this actor is about, as it may be confusing. Having it on a separate page allows people to know at a glance that it is a special user.

It looks like this:
![image](https://user-images.githubusercontent.com/384364/61241511-ed749a00-a743-11e9-876a-da7dd96f00e0.png)

Gargron replied:
Should the `url` maybe just point to the `/about` page then? What I'm trying to avoid is the maintenance cost of another static page design because they've tended to be redesigned quite a few times so far.

ThibG replied:
I don't think so, we really want to explain what this account is for. I should probably reword that description btw.

Gargron commented on config/routes.rb:
```diff
@@ -28,6 +28,10 @@
   get 'intent', to: 'intents#show'
   get 'custom.css', to: 'custom_css#show', as: :custom_css
 
+  resource :instance_actor, path: 'instance-actor', only: [:show] do
```
Maybe just `actor`? There's no overlap either way. Seems redundant.

ThibG replied:
For the controller name? The path? I thought having “instance” in the name would make it clearer that it is meant to represent the instance itself.

Gargron replied:
For the path! Seems cleaner that way for two reasons: Our ActivityPub JSON doesn't use the word "instance" anywhere, and other multi-word URLs use underscores instead of dashes, so there would be a consistency issue.

ThibG replied:
Ok, changed.

Gargron commented on db/migrate/20190715164535_add_instance_actor.rb:
```diff
@@ -0,0 +1,9 @@
+class AddInstanceActor < ActiveRecord::Migration[5.2]
+  def up
+    Account.create!(id: -99, actor_type: 'Application', locked: true, username: Rails.configuration.x.local_domain)
```
You need to add this to `seed.rb` as well for new installations.

ThibG replied:
Oops. I will do that!

Gargron commented:
You should probably make `Account.representative` return this record

ClearlyClaire committed:
Remove unused `icon` and `image` attributes from instance actor

ClearlyClaire committed:
Use if/elsif/else instead of return + ternary operator

ClearlyClaire committed:
Add instance actor to fresh installs

ThibG commented:
> You should probably make `Account.representative` return this record

Yes, I will do that, I just need to think about how to do cleanly, and I'd have to adjust the tests.

ClearlyClaire committed:
Use instance actor as instance representative

Use instance actor for forwarding reports, relay operations, and spam
auto-reporting.

ClearlyClaire committed:
Seed database in test environment

ClearlyClaire committed:
Fix single-user mode

ClearlyClaire committed:
Fix tests

ClearlyClaire committed:
Fix specs to accomodate for an extra `Account`

ClearlyClaire committed:
Auto-reject follows on instance actor

Following an instance actor might make sense, but we are not handling that
right now, so auto-reject.

ClearlyClaire committed:
Fix webfinger lookup and serialization for instance actor

ClearlyClaire committed:
Rename instance actor

ClearlyClaire committed:
Make it clear in the HTML view that the instance actor should not be blocked

ClearlyClaire committed:
Raise cache time for instance actor as there's no dynamic content

ClearlyClaire committed:
Re-use /about/more with a flash message for instance actor profile

ThibG commented:
Changed the instance actor profile URL to `/about/more?instance_actor=true` which displays a flash message:
![image](https://user-images.githubusercontent.com/384364/61494013-5a429b00-a9b5-11e9-8272-dbef14eb5b91.png)

This has the added benefit of always returning json for the actor URI, thus not having to worry about the `Vary` header for caching.

Pizzacus commented:
> It is used for federation purposes and should not be blocked unless you want to block the whole **instance**

So far, the English version of Mastodon has been VERY consistent about using the word "server" instead of "instance".

This would actually be, as far as I know, the first time the word "instance" is used on the actual UI.

This should be addressed should it not?
